### ISSUE TITLE
[BUG] train_batch_size / val_batch_size has a great influence on the result

### ISSUE THREAD

**Issue by JixiangGao** [volcengine/verl#1515]
```#!/bin/bash
set -x
MODEL_PATH=/path/to/Qwen3-8B

export HYDRA_FULL_ERROR=1
export VLLM_ATTENTION_BACKEND=XFORMERS
# ray start --head
python3 -m verl.trainer.main_ppo \
    algorithm.adv_estimator=grpo \
    data.train_files=/path/to/train.parquet \
    data.val_files=/path/to/val.parquet \
    data.train_batch_size=128 \
    data.val_batch_size=512 \
    data.max_prompt_length=2048 \
    data.max_response_length=2048 \
    data.truncation=right \
    actor_rollout_ref.model.path=$MODEL_PATH  \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.actor.ppo_mini_batch_size=16 \
    actor_rollout_ref.actor.use_dynamic_bsz=True \
    actor_rollout_ref.actor.ppo_max_token_len_per_gpu=20480 \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.001 \
    actor_rollout_ref.actor.kl_loss_type=low_var_kl \
    actor_rollout_ref.actor.ulysses_sequence_parallel_size=1 \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.actor.fsdp_config.param_offload=False \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=False \
    actor_rollout_ref.rollout.tensor_model_parallel_size=1 \
    actor_rollout_ref.rollout.name=vllm \
    actor_rollout_ref.rollout.temperature=0.6 \
    actor_rollout_ref.rollout.top_p=0.95 \
    actor_rollout_ref.rollout.top_k=20 \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.85 \
    actor_rollout_ref.rollout.n=8 \
    actor_rollout_ref.rollout.max_num_batched_tokens=32768 \
    actor_rollout_ref.rollout.val_kwargs.temperature=0.6 \
    actor_rollout_ref.rollout.val_kwargs.top_p=0.95 \
    actor_rollout_ref.rollout.val_kwargs.top_k=20 \
    actor_rollout_ref.rollout.val_kwargs.do_sample=True \
    actor_rollout_ref.ref.fsdp_config.param_offload=False \
    algorithm.kl_ctrl.kl_coef=0.001 \
    custom_reward_function.path=/path/to/my_reward.py \
    custom_reward_function.name=compute_score \
    trainer.critic_warmup=0 \
    trainer.logger=['console','wandb','tensorboard'] \
    trainer.project_name='GRPO_exp' \
    trainer.experiment_name='Qwen3-8B-exp' \
    trainer.val_before_train=True \
    trainer.n_gpus_per_node=8 \
    trainer.nnodes=2 \
    trainer.save_freq=100 \
    trainer.test_freq=200 \
    trainer.default_local_dir=/path/to/output \
    trainer.validation_data_dir=/path/to/validation_data_dir \
    trainer.total_epochs=1 $@ 
    
```

When we train the `Qwen3-8b` on 2 nodes using the above command, the training result/output is very poor. However, when we adjust `train_batch_size` to 128 and `val_batch_size` to 16, the training result/output will become very stable. 

The following are the results of training and validating.

![Image](https://github.com/user-attachments/assets/cefb0304-e12f-4e66-99e8-202c86333599)

 
- verl version: the latest commit of master branch.
- vllm: 0.8.4 / 0.8.5post1

**Comment by JixiangGao** [comment]
It is truly normal to train `Deepseek-R1-32B` using a commit from two months ago (the exact one is unknown), but training `Deepseek-R1-32B` on the latest commit also has the same problem as mentioned above.

**Comment by JixiangGao** [comment]
update: `actor_rollout_ref.rollout.name=sglang` works. But `vllm` still does't work.

**Comment by Parasolation2** [comment]
请问这个问题解决了么？我同样的配置在4nodes上val before train的结果比2nodes上val before train的结果稳定低2个点，且response clip ratio会高出很多，4nodes为0.5， 2nodes 为0.25
python3 -m verl.trainer.main_ppo \
    algorithm.adv_estimator=grpo \
    data.train_files= \
    data.val_files= \
    data.train_batch_size=256 \
    data.max_prompt_length=8196 \
    data.max_response_length=8196 \
    ++data.max_valid_prompt_length=8196 \
    data.filter_overlong_prompts=True \
    ++data.val_filter_overlong_prompts=True \
    ++data.val_truncation='error' \
    actor_rollout_ref.model.path=$MODEL_PATH \
    ++actor_rollout_ref.actor.fsdp_config.model_dtype=bfloat16 \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.actor.ppo_mini_batch_size=128 \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=2 \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.001 \
    actor_rollout_ref.actor.kl_loss_type=low_var_kl \
    actor_rollout_ref.actor.entropy_coeff=0 \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.actor.fsdp_config.param_offload=False \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=False \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=4 \
    actor_rollout_ref.rollout.tensor_model_parallel_size=2 \
    actor_rollout_ref.model.trust_remote_code=True \
    actor_rollout_ref.rollout.name=vllm \
    actor_rollout_ref.rollout.max_num_batched_tokens=32768 \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.7 \
    actor_rollout_ref.rollout.n=5 \
    actor_rollout_ref.rollout.temperature=0.6 \
    actor_rollout_ref.rollout.val_kwargs.temperature=0.6 \
    actor_rollout_ref.rollout.val_kwargs.top_p=0.95 \
    actor_rollout_ref.rollout.val_kwargs.n=4 \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=4 \
    actor_rollout_ref.ref.fsdp_config.param_offload=False \
    algorithm.use_kl_in_reward=False \
    trainer.critic_warmup=0 \
    trainer.logger=[console,wandb] \
    trainer.project_name=$project_name \
    trainer.val_before_train=True \
    trainer.experiment_name=$experiment_name \
    trainer.n_gpus_per_node=8 \
    trainer.nnodes=$NODE_COUNT \
    trainer.save_freq=100 \
    trainer.test_freq=10 \
    trainer.rollout_data_dir=$model_output_dir/rollout \
    trainer.validation_data_dir=$model_output_dir/validation \
    trainer.default_local_dir=$model_output_dir/model \
    trainer.total_epochs=20

**Comment by NA-Wen** [comment]
请问这个问题解决了吗，实验中观察到了类似的现象

**Comment by Kirrito-k423** [comment]
中文：
感谢报告。这个问题目前我不建议直接关闭，因为 issue 仍有后续用户反馈类似现象，而且现有代码/PR 证据还不足以证明它已经被修复。

从当前代码看，`data.val_batch_size` 仍会决定 validation dataloader 的 batch size；同时验证阶段会使用 `actor_rollout_ref.rollout.val_kwargs`。默认配置里 validation 是 `do_sample=False`、`temperature=0` 的确定性生成，但本 issue 的命令显式设置了 `actor_rollout_ref.rollout.val_kwargs.do_sample=True`、`temperature=0.6`、`top_p=0.95`、`top_k=20`。因此，`val_batch_size` 改变后影响 vLLM 的 batching/sampling/RNG 行为是一个合理怀疑点，但这还不是可确认的根因。

为了继续定位，能否请提供：
1. 具体 verl commit hash，而不是 “latest master”；
2. 同一 commit 下 `vllm` 与 `sglang` 的最小复现对比；
3. 固定 `actor_rollout_ref.rollout.val_kwargs.do_sample=False`、`temperature=0` 后，`val_batch_size=16` 和 `512` 是否仍有显著差异；
4. 是否设置了 `data.validation_shuffle=False`，以及验证集大小、reward function 是否会受样本顺序或批大小影响；
5. 一小段可公开的 validation outputs/metrics 对比，最好包含 response length、clip ratio、format score。

English:
Thanks for the report. I would not close this yet: there are later comments reporting similar behavior, and I could not find clear evidence that this has already been fixed.

From the current code, `data.val_batch_size` still controls the validation dataloader batch size, and validation generation uses `actor_rollout_ref.rollout.val_kwargs`. The default validation config is deterministic (`do_sample=False`, `temperature=0`), but this issue explicitly sets `actor_rollout_ref.rollout.val_kwargs.do_sample=True`, `temperature=0.6`, `top_p=0.95`, and `top_k=20`. So it is plausible that changing `val_batch_size` changes vLLM batching/sampling/RNG behavior, but that is not yet a confirmed root cause.

Could you please provide:
1. the exact verl commit hash instead of “latest master”;
2. a minimal same-commit comparison between `vllm` and `sglang`;
3. whether the gap still exists with `actor_rollout_ref.rollout.val_kwargs.do_sample=False` and `temperature=0` for both `val_batch_size=16` and `512`;
4. whether `data.validation_shuffle=False` is set, plus validation set size and whether the reward function depends on sample order or batch size;
5. a small public comparison of validation outputs/metrics, ideally including response length, clip ratio, and format score.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
